### PR TITLE
Initial use of GitVersion

### LINKED ISSUE #206: Use GitVersion for the GUI builds
We started to experiment with this in nunit/NUnit.System.Linq#8 but didn't finish it. I'd like to try with the GUI and then see if we can push the approach back to other NUnit projects.

The general idea is that we would have __no__ text in our source or scripts that specified the version. GitVersion would figure it all out. This is a big step toward automatic releasing.

### PR DESCRIPTION
**Pull request by CharliePoole**
Fixes #206

This is initial implementation of the use of GitVersion to determine the correct version for each build. When complete, it will no longer be necessary to track and update the version of the GUI in the source code or scripts.

The initial commit calculates and displays the version info that __would be used__ but doesn't actually use it. By creating a PR, I'll be able to see how the pre-release tag portion of the version is set for PRs, but the PR should not be merged by anyone but me, after it's working correctly.

Please review and comment on the versioning scheme anyway!

### CODE REVIEW

File: build.cake [hunk 1]
```diff
@@ -2,6 +2,8 @@
 #tool nuget:?package=GitVersion.CommandLine
 #addin "Cake.Incubator"
 
+using System.Text.RegularExpressions;
```
Comment (mikkelbu) [inline build.cake]: Why is this necessary ?

File: build.cake [hunk 2]
```diff
@@ -1,4 +1,8 @@
 #tool nuget:?package=NUnit.ConsoleRunner&version=3.6.1
```
Comment (mikkelbu) [inline build.cake]: We are back to using both tabs and spaces for indention in this file.

File: build.cake (hunk 1 above)
Comment (CharliePoole) [reply build.cake]: I had a regular expression at one point but it's gone now.

File: build.cake (hunk 2 above)
Comment (CharliePoole) [reply build.cake]: Yes. I should try to tell VS how to handle this file.

**Comment by CharliePoole** [comment]
@NUnitSoftware/gui-team @ChrisMaddock @rprouse
I think this is in it's final form as far as coding. If you reviewed it earlier, it may be worth another look.

That said, it's still not doing anything but merely displaying what it __would__ do. Next commits will actually begin to impact the packaging.

I'm saving modification of the AssemblyInfo for last. 😃

**Comment by immeraufdemhund** [comment]
pardon if this is a dumb question but is this changing how we do our branching and do our pull requests??? or just how the CI's handle our PR's and branches?

**Comment by CharliePoole** [comment]
It's changing how we version. Normally, when you do a branch and PR, you don't set the version. Our script handles it. But the script depends on the version being coded in it somewhere - as well as in the AssemblyInfo files.

Once we implement this, I'd hope to end up where we no longer have the version in any of our source code and therefore don't need to update it. GitVersion can examine the repository history and decide what version we want, based on some conventions and the settings we establish.

Currently, we create automatic pre-release suffixes for our versions on AppVeyor only, using environment variables they provide. Once this is in place, we'll have a consistent automatic method that works on AppVeyor, Travis and locally. 

Initially, final (non-preview) releases will continue to be done manually. But this will enable a later change to do them automatically as well. All that adds up to more frequent automated releases, which has to be a good thing!

**Comment by CharliePoole** [comment]
With the latest commits, we are now using the info from GitVersion to do packaging. The hard-coded version in the script is no longer used.

There is still quite a lot of debug info in the script, which I will remove when it's no longer needed. I'll reformat using spaces everywhere before merge as well.

Note that we can't do a chocolatey package on Linux without a bunch of work, so I'm simply not creating one.

Next step is to deal with the AssemblyInfo files.

**Comment by CharliePoole** [comment]
This is ready for review and merge once CI passes.

**Comment by mikkelbu** [comment]
I'll take a look at it tonight (in approx. 3 hours)

File: .vscode/tasks.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
+ {
+    // See https://go.microsoft.com/fwlink/?LinkId=733558
+    // for the documentation about the tasks.json format
+    "version": "2.0.0",
```
[the remark this answers is not in the document]
Comment (CharliePoole) [reply .vscode/tasks.json]: That's because I executed tasks in the script under VsCode. File should be in .gitignore.

File: build.cake [hunk 4]
```diff
@@ -1,4 +1,8 @@
...
+#tool nuget:?package=GitVersion.CommandLine
+#addin "Cake.Incubator"
+
+using System.Text.RegularExpressions;
```
[the remark this answers is not in the document]
Comment (CharliePoole) [reply build.cake]: Exactly.

File: build.cake [hunk 5]
```diff
@@ -167,9 +144,7 @@ Task("PackageZip")
             BIN_DIR + "CHANGES.txt",
             BIN_DIR + "nunit-gui.exe",
             BIN_DIR + "nunit-gui.exe.config",
-            BIN_DIR + "nunit-gui.pdb",
```
[the remark this answers is not in the document]
Comment (CharliePoole) [reply build.cake]: They break the build under Mono 4.6.2, which doesn't produce them. We could put them back if we used logic to include them only when they actually are present.

File: .vscode/tasks.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (CharliePoole) [reply .vscode/tasks.json]: The script is build.cake, Package is a target.

File: build.cake (hunk 5 above)
[the remark this answers is not in the document]
Comment (mikkelbu) [reply build.cake]: Ok. Let us just leave them out to begin with. Then we can always include them later.

File: .vscode/tasks.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikkelbu) [reply .vscode/tasks.json]: Ok. The ` "script": "Package"` was a bit misleading 😄

File: .travis.yml [hunk 6]
```diff
@@ -4,5 +4,5 @@ mono:
   - latest
   - 4.6.2
 script: 
-  - ./build.sh -target="Travis" -configuration="Release" -verbosity="verbose"
-    
+  - git fetch --unshallow
+  - ./build.sh --target "Travis"
```
Comment (mikkelbu) [inline .travis.yml]: I've pushed a commit b758c79 to correct the error from the merge (the error made Travis fail every time).

**Comment by mikkelbu** [comment]
@CharliePoole .tavis.yml should contain something like 
`  - ./build.sh -target="Travis" -configuration="Release" -verbosity="verbose"`

**Comment by CharliePoole** [comment]
Can you explain further? Is the problem the dash, the missing config, the need for verbosity? Is this because of merging the new Cake bootstrap script? Seems like, if it messes us up, we might not want it.

**Comment by CharliePoole** [comment]
Going to bed... I'll look at this in the morning.

**Comment by mikkelbu** [comment]
When you merged master to this branch no change was made to .tavis.yml (as far as I can tell), even though it was changed on master (I don't know much about GIT merges, but I guess that there must have been a conflict since you had made changes to the file on this branch). So we lost the changes done to .tavis.yml. So I made the commit https://github.com/NUnitSoftware/nunit-gui/pull/226/commits/b758c793719bb8d05c820877c79025351d0a9c2f (on this branch) to solve the problem, but I think that you concurrently also tried to solve the problem (see ae9e7a3dc75406ef) and as a result overwrote my change.

 .tavis.yml on master
`- ./build.sh -target="Travis" -configuration="Release" -verbosity="verbose"`
 .tavis.yml on this branch 
`- ./build.sh -t "Travis"`
So `-t ` should be replaced with `-target=`. The part about configuration and verbosity is just to keep the old behaviour as these are not the default anymore.

You can also see the diff between master and this branch here
https://github.com/NUnitSoftware/nunit-gui/pull/226/files#diff-354f30a63fb0907d4ad57269548329e3

And yes, it is because of the new bootstrap script, but I don't think that it is more problematic than other changes we do (and since we change .travis.yml quite rarely I don't think that it will effect us much).

**Comment by CharliePoole** [comment]
I understand that it would probably work if I put it back to how it was in master. It would also probably work if I backed out the upgraded build.sh script. I'm trying to figure out which is the best course of action.

I take it you are saying you merely used the args from master, and they work. But those args are wrong in several senses. We don't want verbose output. We shouldn't have to specify the configuration, since it is supposed to use our default in `build.cake` - although specifying it here is no big deal.

@jnm2 Is this related to the changes you made to Cake? If so can you give us info about exactly what changed in the command-line args or point to same?

**Comment by jnm2** [comment]
@CharliePoole I put all the details in https://github.com/cake-build/resources/issues/33. I was just on Gitter bugging them again to get https://github.com/cake-build/resources/pull/34 validated and merged.

**Comment by CharliePoole** [comment]
@jnm2 The command that was failing for me most recently was

```
./build.sh -t "Travis"
```

before that this was failing:

```
./build.sh --target "Travis"
```

Is it the quotation marks?

**Comment by jnm2** [comment]
@CharliePoole nothign has changed WRT quotation marks, thankfully! The issue is that right now the *only* format build.sh understands is the format Cake.exe understands, which is `--target="Foo"` and `--target=Foo`. Note the `=`.

Hopefully the PR is merged soon and you can go back to separating with a space and `-t`. :-/

**Comment by CharliePoole** [comment]
It's the `=`! Now I get it. 😀

**Comment by mikkelbu** [comment]
> I take it you are saying you merely used the args from master, and they work. But those args are wrong in several senses. We don't want verbose output. We shouldn't have to specify the configuration, since it is supposed to use our default in build.cake - although specifying it here is no big deal.

The reason for why I chose set `-configuration="Release" -verbosity="verbose"` was to keep the old behaviour. The old version of build.sh contained defaults which set configuration=Release and verbosity=verbose, if no arguments were given to build.sh (and we only supplied target).
(see e.g. here https://github.com/NUnitSoftware/nunit-gui/blob/a78c5bf8d240214b55283e25edc5b11f139b4313/build.sh#L13-L19)

**Comment by jnm2** [comment]
This is the best place to keep defaults:
https://github.com/NUnitSoftware/nunit-gui/blob/e0699e256f7ce72b23c4cb2561122f2ce7199654/build.cake#L10-L11

Cake's template has `Release` which means someone changed it, even though the change never worked due to the design fault in the the bootstrappers. If you ever customize these, rather than editing in each bootstrapper, it would seem to make the most sense to put your argument defaults in the build script.
Not in the bootstrappers and not in CI server config files. I'd only put argument *overrides* in CI server configs where they are relevant and keep defaults in build.cake where they are most relevant.

**Comment by CharliePoole** [comment]
@mikkelbu Yes, I now understand that. It's a breaking change on the part of the Cake folks. I agree with how you compensated and put back in the configuration arg. Problem is, these updates come out without any release info on the part of Cake.

@jnm2 Wrong. 😄 We got Debug as a default when we ran Cake directly and Release when we ran build.ps1 or build.sh. I worked for quite a while using Cake directly locally and only using the bootstrap scripts in CI. I always thought it was weird that we defaulted to Release in one case and Debug in the other, but I got used to it and our CI scripts came to depend on it.

I agree it makes the most sense for the bootstrap scripts to be neutral in terms of defaults that are passed to Cake. It's just a change from the past and not one that's easy to notice in all cases.

**Comment by jnm2** [comment]
My bad. Thanks for filling me in on the history behind it.
